Cannot parse expression segment

I tried this code:

```rust
trait Foo { }

trait Bar {
    type Item: Foo;
}

impl Foo for i32 { }
impl Bar for i32 {
    type Item = i32;
}

fn only_foo<T: Foo>() { }

fn only_bar<T: Bar>() {
    // `T` implements `Bar` hence `<T as Bar>::Item` must also implement `Bar`
    only_foo::<T::Item>()
}

fn main() {
    only_bar::<i32>();
    only_foo::<<i32 as Bar>::Item>();
}
```

I expected to see this happen: Compiles correctly

Instead, this happened:

```
./test.rs:24:13: error: could not parse path expression segment
   24 |     only_foo::<<i32 as Bar>::Item>();
      |             ^~
./test.rs:24:28: error: found null denotation scope resolution operator, and have not written handling for it
   24 |     only_foo::<<i32 as Bar>::Item>();
      |                            ^~
./test.rs:24:5: error: failed to parse statement or expression in block expression
   24 |     only_foo::<<i32 as Bar>::Item>();
      |     ^~~~~~~~
./test.rs:24:30: error: unrecognised token ‘identifier’ for start of item
   24 |     only_foo::<<i32 as Bar>::Item>();
      |                              ^~~~
./test.rs:24:30: error: failed to parse item in crate
```

### Meta

https://github.com/Rust-GCC/gccrs/commit/007166ca6ceecd429f7033336af25dd679424ca2

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -6804,11 +6804,13 @@ Parser<ManagedTokenSource>::parse_path_expr_segment ()
   /* use lookahead to determine if they actually exist (don't want to
    * accidently parse over next ident segment) */
   if (lexer.peek_token ()->get_id () == SCOPE_RESOLUTION
-      && lexer.peek_token (1)->get_id () == LEFT_ANGLE)
+      && (lexer.peek_token (1)->get_id () == LEFT_ANGLE
+	  || lexer.peek_token (1)->get_id () == LEFT_SHIFT))
     {
       // skip scope resolution
       lexer.skip_token ();
 
+      // Let parse_path_generic_args split "<<" tokens
       AST::GenericArgs generic_args = parse_path_generic_args ();
 
       return AST::PathExprSegment (std::move (ident), locus,

--- a/gcc/testsuite/rust/compile/parse_generic_path_expr.rs
+++ b/gcc/testsuite/rust/compile/parse_generic_path_expr.rs
@@ -1,0 +1,4 @@
+// { dg-additional-options "-frust-compile-until=ast" }
+fn main() {
+    only_foo::<<i32 as Bar>::Item>();
+}

